components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java

```diff
@@ -170,32 +180,66 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
-                    } else {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Scope validation failed from "+ validator);
+                try {
+                    if (validator != null && appScopeValidators.contains(validator.getValidatorName())) {
+                        //take the intersection of defined scope validators and scope
+                        //validators registered for the application
+                        log.debug(String.format("Validating scope of token %s using %s", accessTokenDO.getTokenId(),
+                                validator.getValidatorName()));
+
+                        boolean isValid = validator.validateScope(accessTokenDO, resource);
+                        if (!isValid) {
+                            log.debug(String.format("Scope validation failed for %s", validator.getValidatorName()));
+                            apiKeyValidationInfoDTO.setAuthorized(false);
+                            apiKeyValidationInfoDTO.setValidationStatus(APIConstants.KeyValidationStatus.INVALID_SCOPE);
+                            return false;
                         }
-                        apiKeyValidationInfoDTO.setAuthorized(false);
-                        apiKeyValidationInfoDTO.setValidationStatus(APIConstants.KeyValidationStatus.INVALID_SCOPE);
-                        return false;
+                        appScopeValidators.remove(validator.getValidatorName());
                     }
+                } catch (IdentityOAuth2Exception e) {
+                    log.error("ERROR while validating token scope " + e.getMessage(), e);
+                    apiKeyValidationInfoDTO.setAuthorized(false);
+                    apiKeyValidationInfoDTO.setValidationStatus(APIConstants.KeyValidationStatus.INVALID_SCOPE);
+                    return false;
                 }
             }
+
+            if (!appScopeValidators.isEmpty()) {   //if scope validators are not defined in identity.xml but there are scope validators assigned to an application, throws exception.
+                throw new IdentityOAuth2Exception(String.format("The scope validator(s) %s registered for application %s@%s" +
+                                " is/are not found in the server configuration ", StringUtils.join(appScopeValidators, ", "),
+                        appInfo.getApplicationName(), OAuth2Util.getTenantDomainOfOauthApp(appInfo)));
+            }
+
+        } catch (InvalidOAuthClientException e) {
+            log.error("Could not Fetch Application data for client with clientId = " + clientId);
+            apiKeyValidationInfoDTO.setAuthorized(false);
+            apiKeyValidationInfoDTO.setValidationStatus(APIConstants.KeyValidationStatus.INVALID_SCOPE);
+            return false;
         } catch (IdentityOAuth2Exception e) {
-            log.error("ERROR while validating token scope " + e.getMessage(), e);
+            log.error("Error while retrieving the app information");
```

Comment:
which application can be add here. so it will be easy to identify issues.

Reply:
Fixed with b58b5414eb4051494cf7768e5ebfea2d5d7a5933